examples/generative/stylegan.py

```diff
@@ -0,0 +1,736 @@
+from tensorflow.keras.layers import UpSampling2D, AveragePooling2D, Flatten
+from tensorflow.keras.models import Sequential, load_model
+from tensorflow.keras.callbacks import ModelCheckpoint
+from tensorflow_addons.layers import InstanceNormalization
+
+
+"""
+## Prepare Dataset
+In this examples, we wiil use CelebA pre-resized to 128x128 resolution.
+"""
+
+CKPT_PATH = "./ckpt/64x64"
+DATA_PATH = "celeba_64x64/*.jpg"
+RESULT_PATH = "./results"
+os.makedirs(CKPT_PATH, exist_ok=True)
+os.makedirs(RESULT_PATH, exist_ok=True)
+
+"""shell
+curl -LO
+https://github.com/soon-yau/stylegan_keras/releases/download/keras_example/celeba_64x64.zip
+https://github.com/soon-yau/stylegan_keras/releases/download/keras_example/celeba_64x64.zip
+unzip celeba_64x64.zip -d .
+"""
+
+"""
+We use different batch size for different image resolution, with small batch size for
+high resolution so they fit into GPU memory. The training steps are scaled accordingly.
+"""
+
+BATCH_SIZE = {2: 16, 3: 16, 4: 16, 5: 16, 6: 16, 7: 8, 8: 4, 9: 4, 10: 4}
+TRAIN_STEP_RATIO = {k: BATCH_SIZE[2] / v for k, v in BATCH_SIZE.items()}
+
+
+def load(res, image_file):
+    image = tf.io.read_file(image_file)
+    image = tf.image.decode_jpeg(image)
+    image = tf.image.resize(
+        image, [res, res], method=tf.image.ResizeMethod.NEAREST_NEIGHBOR
+    )
+    image = tf.cast(image, tf.float32)
+    image = (image / 127.5) - 1
+    return image
+
+
+def create_dataset_(data_path, res):
+    res_log2 = int(np.log2(res))
+    BUFFER_SIZE = 100
+    images = glob(data_path)
+    random.shuffle(images)
+    batch_size = BATCH_SIZE[res_log2]
+    tf_list = tf.data.Dataset.from_tensor_slices(images)
+    dataset = tf_list.map(
+        partial(load, res), num_parallel_calls=tf.data.experimental.AUTOTUNE
+    )
+    dataset = (
+        dataset.shuffle(BUFFER_SIZE).batch(batch_size, drop_remainder=True).repeat()
+    )
+    return dataset
+
+
+create_dataset = partial(create_dataset_, DATA_PATH)
+
+
+"""
+### Utility function to display images after each epoch
+"""
+
+
+def plot_images(images, log2_res, fname=""):
+    scales = {2: 0.5, 3: 1, 4: 2, 5: 3, 6: 4, 7: 5, 8: 6, 9: 7, 10: 8}
+    scale = 0.5 * scales[log2_res]
+
+    grid_col = min(images.shape[0], int(16 // scale))
+    grid_row = 1
+
+    f, axarr = plt.subplots(
+        grid_row, grid_col, figsize=(grid_col * scale, grid_row * scale)
+    )
+
+    for row in range(grid_row):
+        ax = axarr if grid_row == 1 else axarr[row]
+        for col in range(grid_col):
+            ax[col].imshow(images[row * grid_col + col])
+            ax[col].axis("off")
+    plt.show()
+    if fname:
+        f.savefig(fname)
+
+
+"""
+## Custom Layers
+The following are building blocks that will be used to construct the generators and
+discriminators.
+"""
+
+
+def fade_in(alpha, a, b):
+
+    return alpha * a + (1.0 - alpha) * b
+
+
+def wasserstein_loss(y_true, y_pred):
+
+    return -tf.reduce_mean(y_true * y_pred)
+
+
+class PixelNorm(Layer):
```

Comment:
This should be a single-line function, not a layer.

```python
def pixel_norm(x, epsilon=1e-8):
    return x / tf.math.sqrt(tf.reduce_mean(x ** 2, axis=-1, keepdims=True) + epsilon)
```

Reply:
The reason I made this into a layer is so that this can be easily swapped with other (batch) normalization layer if needed. Since this will not change, ok, I'll convert that into a function.